src/ripple/net/ShardDownloader.md

```diff
@@ -0,0 +1,263 @@
+# Shard Downloader Process
+
+## Overview
+
+This document describes mechanics of the `SSLHTTPDownloader`, a class that performs the task of downloading shards from remote web servers via
+SSL HTTP. The downloader utilizes a strand (`boost::asio::io_service::strand`) to ensure that downloads are never executed concurrently. Hence, if a download is in progress when another download is initiated, the second download will be queued and invoked only when the first download is completed.
+
+## New Features
+
+The downloader has been recently (March 2020) been modified to provide some key features:
+
+- The ability to stop downloads during a graceful shutdown.
+- The ability to resume partial downloads after a crash or shutdown.
+- <span style="color:gray">*(Deferred) The ability to download from multiple servers to a single file.*</span>
+
+## Classes
+
+Much of the shard downloading process concerns the following classes:
+
+- `SSLHTTPDownloader`
+
+   This is a generic class designed for serially executing downloads via HTTP SSL.
+
+- `ShardArchiveHandler`
+
+   This class uses the `SSLHTTPDownloader` to fetch shards from remote web servers. Additionally, the archive handler performs sanity checks on the downloaded files and imports the validated files into the local shard store.
+
+   The `ShardArchiveHandler` exposes a simple public interface:
+
+   ```C++
+   /** Add an archive to be downloaded and imported.
+   @param shardIndex the index of the shard to be imported.
+   @param url the location of the archive.
+   @return `true` if successfully added.
+   @note Returns false if called while downloading.
+   */
+   bool
+   add(std::uint32_t shardIndex, std::pair<parsedURL, std::string>&& url);
+
+   /** Starts downloading and importing archives. */
+   bool
+   start();
+   ```
+
+   When a client submits a `download_shard` command via the RPC interface, each of the requested files is registered with the handler via the `add` method. After all the files have been registered, the handler's `start` method is invoked, which in turn creates an instance of the `SSLHTTPDownloader` and begins the first download. When the download is completed, the downloader invokes the handler's `complete` method, which will initiate the download of the next file, or simply return if there are no more downloads to process. When `complete` is invoked with no remaining files to be downloaded, the handler and downloader  are not destroyed automatically, but persist for the duration of the application.
```

Comment:
No, originally they would get destroyed once all downloads were completed and subsequent downloads would be facilitated by new instances of these classes. I wanted to preserve this feature but it would be require significant modifications to the `Stoppable` interface. So the only reason they persist is to comply with graceful shutdowns via `Stoppable`. Do you think I should say that in the doc?

Reply:
Yeah, I think it's a good idea to say that they are persisted to assist with graceful shutdowns by `Stoppable`.  That way folks will understand the motivation.